docs/logging_snippets.py

```diff
@@ -0,0 +1,369 @@
+    grants = dataset.access_grants
+    grants.append(AccessGrant(
+        'WRITER', 'groupByEmail', 'cloud-logs@google.com'))
+    dataset.access_grants = grants
+    dataset.update()
+    # [END sink_dataset_permissions]
+
+    return dataset
+
+
+@snippet
+def sink_bigquery(client, to_delete):
+    """Sink log entries to bigquery."""
+    dataset = _sink_bigquery_setup(client)
+    to_delete.append(dataset)
+    SINK_NAME = 'robots-bigquery-%d' % (_millis(),)
+    FILTER = 'textPayload:robot'
+
+    # [START sink_bigquery_create]
+    DESTINATION = 'bigquery.googleapis.com%s' % (dataset.path)
+    sink = client.sink(SINK_NAME, filter_=FILTER, destination=DESTINATION)
+    assert not sink.exists()  # API call
+    sink.create()  # API call
+    assert sink.exists()  # API call
+    # [END sink_bigquery_create]
+    to_delete.insert(0, sink)  # delete sink before dataset
+
+
+def _sink_pubsub_setup(client):
+    from google.cloud import pubsub
+    TOPIC_NAME = 'sink-pubsub-%d' % (_millis(),)
+    client = pubsub.Client()
+    topic = client.topic(TOPIC_NAME)
+    topic.create()
+
+    # [START sink_topic_permissions]
+    policy = topic.get_iam_policy()
+    policy.owners.add(policy.group('cloud-logs@google.com'))
+    topic.set_iam_policy(policy)
+    # [END sink_topic_permissions]
+
+    return topic
+
+
+@snippet
+def sink_pubsub(client, to_delete):
+    """Sink log entries to pubsub."""
+    topic = _sink_pubsub_setup(client)
+    to_delete.append(topic)
+    SINK_NAME = 'robots-pubsub-%d' % (_millis(),)
+    FILTER = 'logName:apache-access AND textPayload:robot'
+    UPDATED_FILTER = 'textPayload:robot'
+
+    # [START sink_pubsub_create]
+    DESTINATION = 'pubsub.googleapis.com/%s' % (topic.full_name)
+    sink = client.sink(SINK_NAME, filter_=FILTER, destination=DESTINATION)
+    assert not sink.exists()  # API call
+    sink.create()  # API call
+    assert sink.exists()  # API call
+    # [END sink_pubsub_create]
+    to_delete.insert(0, sink)  # delete sink before topic
+
+    # [START client_list_sinks]
+    for sink in client.list_sinks():  # API call(s)
+        do_something_with(sink)
+    # [END client_list_sinks]
+
+    # [START sink_reload]
+    existing_sink = client.sink(SINK_NAME)
+    existing_sink.reload()
+    # [END sink_reload]
+    assert existing_sink.filter_ == FILTER
+    assert existing_sink.destination == DESTINATION
+
+    # [START sink_update]
+    existing_sink.filter_ = UPDATED_FILTER
+    existing_sink.update()
+    # [END sink_update]
+    existing_sink.reload()
+    assert existing_sink.filter_ == UPDATED_FILTER
+
+    # [START sink_delete]
+    sink.delete()
+    # [END sink_delete]
+    to_delete.pop(0)
+
+
+def _line_no(func):
+    code = getattr(func, '__code__', None) or getattr(func, 'func_code')
+    return code.co_firstlineno
+
+
+def _find_examples():
+    funcs = [obj for obj in globals().values()
+             if getattr(obj, '_snippet', False)]
+    for func in sorted(funcs, key=_line_no):
```

Comment:
Because "source order" groups the tests "logically", rather than in random hash order.